core/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java

```diff
@@ -914,55 +916,69 @@ public MetaData build() {
+                        }
+                    }
+                }
+                assert duplicates.size() > 0;
+                throw new IllegalStateException("index and alias names need to be unique, but the following duplicates were found ["
+                    + Strings.join(duplicates, ", ")+ "]");
+
             }
-            String[] allOpenIndices = allOpenIndicesLst.toArray(new String[allOpenIndicesLst.size()]);
-            String[] allClosedIndices = allClosedIndicesLst.toArray(new String[allClosedIndicesLst.size()]);
 
             // build all indices map
             SortedMap<String, AliasOrIndex> aliasAndIndexLookup = new TreeMap<>();
             for (ObjectCursor<IndexMetaData> cursor : indices.values()) {
                 IndexMetaData indexMetaData = cursor.value;
-                aliasAndIndexLookup.put(indexMetaData.getIndex().getName(), new AliasOrIndex.Index(indexMetaData));
+                AliasOrIndex exiting = aliasAndIndexLookup.put(indexMetaData.getIndex().getName(), new AliasOrIndex.Index(indexMetaData));
+                assert exiting == null : "duplicate for " + indexMetaData.getIndex();
 
                 for (ObjectObjectCursor<String, AliasMetaData> aliasCursor : indexMetaData.getAliases()) {
                     AliasMetaData aliasMetaData = aliasCursor.value;
-                    AliasOrIndex aliasOrIndex = aliasAndIndexLookup.get(aliasMetaData.getAlias());
-                    if (aliasOrIndex == null) {
-                        aliasOrIndex = new AliasOrIndex.Alias(aliasMetaData, indexMetaData);
-                        aliasAndIndexLookup.put(aliasMetaData.getAlias(), aliasOrIndex);
-                    } else if (aliasOrIndex instanceof AliasOrIndex.Alias) {
-                        AliasOrIndex.Alias alias = (AliasOrIndex.Alias) aliasOrIndex;
-                        alias.addIndex(indexMetaData);
-                    } else if (aliasOrIndex instanceof AliasOrIndex.Index) {
-                        AliasOrIndex.Index index = (AliasOrIndex.Index) aliasOrIndex;
-                        throw new IllegalStateException("index and alias names need to be unique, but alias [" + aliasMetaData.getAlias() + "] and index " + index.getIndex().getIndex() + " have the same name");
+                    AliasOrIndex alias = aliasAndIndexLookup.get(aliasMetaData.getAlias());
+                    if (alias == null) {
+                        alias = new AliasOrIndex.Alias(aliasMetaData, indexMetaData);
+                        aliasAndIndexLookup.put(aliasMetaData.getAlias(), alias);
                     } else {
-                        throw new IllegalStateException("unexpected alias [" + aliasMetaData.getAlias() + "][" + aliasOrIndex + "]");
+                        assert alias instanceof AliasOrIndex.Alias : alias.getClass().getName();
+                        ((AliasOrIndex.Alias) alias).addIndex(indexMetaData);
```

Comment:
I think this would be cleaner as

```java
aliasAndIndexLookup.compute(aliasMetaData.getAlias(), (aliasName, alias) -> {
    if (alias == null) {
        return new AliasOrIndex.Alias(aliasMetaData, indexMetaData);
    } else {
        ((AliasOrIndex.Alias) alias).addIndex(indexMetaData);
        return alias;
    }
});
```

Reply:
sure